### PR TITLE
Pivot Exports/Downloads Should Not include pivot-grouping or 'Extra' Rows

### LINKED ISSUE #46561: Exports of Pivot Table contains pivot-grouping on version 0.50.15
### Describe the bug

When downloading results from a pivot table (.csv, .xlsx or .json) the results are coming with a `pivot-grouping` column which is not available in the question. The results do not match what the question is displaying.

### To Reproduce

1. Go to any pivot table question in a dashboard or the question itself
2. Click on "Download results" (*if on dashboard*) or "Download full results" (*in the question*)
3. Select the download format (*.csv, .xlsx, or .json*) and download the results
4. Open the downloaded file, there will be a `pivot-grouping` column in it

### Expected behavior

Don't expose system data like pivot-grouping to users.

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.23+9",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.23",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.23+9",
    "os.name": "Linux",
    "os.version": "5.15.0-1026-aws",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "snowflake",
      "h2",
      "mysql"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "13.13"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    },
    "run-mode": "prod",
    "plan-alias": "",
    "version": {
      "date": "2024-07-23",
      "tag": "v0.50.15",
      "hash": "c6697cf"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

blocking some users

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by adam-james-v**
Fixes #46561

When downloading a pivot table's data, a column 'pivot-grouping' is included as well as 'extra' rows that correspond to various totals in the table.

This PR removes the pivot-grouping column and any of these 'extra' rows from the downloads/exports.

### CODE REVIEW

File: test/metabase/api/downloads_exports_test.clj [hunk 1]
```diff
@@ -42,12 +42,21 @@
                  (->>  (spreadsheet/cell-seq r)
                        (mapv read-cell-with-formatting)))))))
 
+(defn- read-json
+  [result]
+  (let [ks (keys (first result))
+        header (map name ks)]
+    (concat
+     [(vec header)]
+     (map (fn [row] (mapv #(get row %) ks)) result))))
```
Comment (metamben) [inline test/metabase/api/downloads_exports_test.clj:51]:
```suggestion
  (let [ks (keys (first result))]
    (cons
     (mapv name ks)
     (map #(mapv % ks) result))
```
It's not needed right now, but handling empty results could be a good idea.

File: test/metabase/api/downloads_exports_test.clj [hunk 2]
```diff
@@ -42,12 +42,21 @@
                  (->>  (spreadsheet/cell-seq r)
                        (mapv read-cell-with-formatting)))))))
 
+(defn- read-json
```
Comment (metamben) [inline test/metabase/api/downloads_exports_test.clj:45]: This is an unusual name. Should it be `tabulate-maps` or something similar?

**Review by metamben — APPROVED** [review]
There are some parts that I found difficult to follow, some comments would be helpful.

File: src/metabase/query_processor/streaming/csv.clj [hunk 3]
```diff
@@ -90,25 +94,33 @@
                      (mapv #(formatter/create-formatter results_timezone % viz-settings) ordered-cols)
                      (vec (repeat (count ordered-cols) identity))))
           ;; write the column names for non-pivot tables
-          (when col-names
-            (write-csv writer [col-names])
+          (when-not opts
+            (let [modified-row (m/remove-nth (or @pivot-grouping-idx (inc (count col-names))) col-names)]
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply src/metabase/query_processor/streaming/csv.clj:98]: Ah, this was a bit of code I wrote quickly and did not come back to clean up. You're right that it's just a 'funny no op', so I'll use your much cleaner suggestion, thanks!

File: src/metabase/query_processor/streaming/csv.clj [hunk 4]
```diff
@@ -73,15 +74,18 @@
...
     (reify qp.si/StreamingResultsWriter
       (begin! [_ {{:keys [ordered-cols results_timezone format-rows? pivot-export-options]
                    :or   {format-rows? true}} :data} viz-settings]
-        (let [opts      (when (and *pivot-export-post-processing-enabled* pivot-export-options)
-                          (assoc pivot-export-options :column-titles (mapv :display_name ordered-cols)))
+        (let [opts           (when (and *pivot-export-post-processing-enabled* pivot-export-options)
+                               (assoc pivot-export-options :column-titles (mapv :display_name ordered-cols)))
               ;; col-names are created later when exporting a pivot table, so only create them if there are no pivot options
-              col-names (when-not opts (common/column-titles ordered-cols (::mb.viz/column-settings viz-settings) format-rows?))]
+              col-names      (common/column-titles ordered-cols (::mb.viz/column-settings viz-settings) format-rows?)
```
[the remark this answers is not in the document]
Comment (adam-james-v) [reply src/metabase/query_processor/streaming/csv.clj:85]: turns out it wasn't a necessary change

**Comment by natashamathur** [comment]
Which release will this come out in?

**Comment by piranha** [comment]
@natashamathur that's going to be the first release in the 50.* after the backport merges, so something like 50.27 or 28.

**Comment by synque-dev-admin** [comment]
@piranha , please let us know when is 50.27 will be released. Please and thank you!

**Comment by piranha** [comment]
@synque-dev-admin I don't have any good tools to do that, so in case I'm not very on time: add yourself to watch backport, and when it's merged next release will contain that.

**Comment by piranha** [comment]
@natashamathur @synque-dev-admin 50.27 is out :)

**Comment by natashamathur** [comment]
Thank you!

**Comment by synque-dev-admin** [comment]
Thank you @piranha

**Pull request merged**